### PR TITLE
iterate: new template function

### PR DESCRIPTION
**Pull request by furiel**
This patchset adds a new template function: `iterate`.

You can think of `iterate` as a generator. The idea is the `iterate` function receives an initial value `x`, and an `f` function, and stores a current value in `$_`. When evaluated, `iterate` prints the current value, and also updates current value to `f(it)`. In other words, through evaluation, `iterate` prints `x`, `f(x)`, `f(f(x))`, ...

```
$(iterate $(+ 1 $_) 0)
```

With that, you can implement interesting stuff with syslog-ng, for example a sequence number.
Or even you can use this to loop through a list.

I would like to use these for tests: I want to generate different messages with `example-msg-generator`.

# Examples

## Sequence number
```
@version: 3.26

log {
    source { example-msg-generator(num(3) freq(0.001) template("$(iterate $(+ 1 $_) 0)")); };
    destination { file(/dev/stdout); };
};
```
will print
```
$ timeout 1 ./syslog-ng -Fe -f ../etc/iterate.conf 
[2020-03-29T11:54:13.488177] syslog-ng starting up; version='3.26.1.140.g2ff84d4'
Mar 29 11:54:13 furiel 0
Mar 29 11:54:13 furiel 1
Mar 29 11:54:13 furiel 2
[2020-03-29T11:54:14.460958] syslog-ng shutting down; version='3.26.1.140.g2ff84d4'
$ 
```

## Loop through list

```
@version: 3.26

log {
    source { example-msg-generator(num(6) freq(0.001)
             template("$(list-nth
                           $(iterate
                               $(% $(+ 1 $_) 3) 0)
                               \"msg1\",\"msg2\",\"msg3\")")); };
    destination { file(/dev/stdout); };
};
```

results in

```
$ timeout 1 ./syslog-ng -Fe -f ../etc/iterate.conf 
[2020-03-29T11:57:08.561749] syslog-ng starting up; version='3.26.1.140.g2ff84d4'
Mar 29 11:57:08 furiel msg1
Mar 29 11:57:08 furiel msg2
Mar 29 11:57:08 furiel msg3
Mar 29 11:57:08 furiel msg1
Mar 29 11:57:08 furiel msg2
Mar 29 11:57:08 furiel msg3
[2020-03-29T11:57:09.540061] syslog-ng shutting down; version='3.26.1.140.g2ff84d4'
$ 
```

### CODE REVIEW

**Comment by kira-syslogng** [comment]
Build SUCCESS

**Comment by furiel** [comment]
As Bazsi mentioned in https://github.com/syslog-ng/syslog-ng/issues/2926#issuecomment-605644168, there is `$_` macro, that I can use for referencing the list elements. I replaced `it` to that. As `$_` is pretty much standard in syslog-ng, I do not feel the necessity to add the anaphoric qualifier to the name. So I am staying with the `iterate` name.

Removed the WIP flag, ready for review.

**Comment by kira-syslogng** [comment]
Build SUCCESS

File: modules/basicfuncs/tf-iterate.c [hunk 1]
```diff
@@ -0,0 +1,107 @@
...
+
+static void
+tf_iterate_free_state(gpointer s)
+{
+  IterateState *state = (IterateState *)s;
+
+  log_template_unref(state->template);
+  state->template = NULL;
+  g_string_free(state->current, TRUE);
+  state->current = NULL;
```
Comment (MrAnno) [inline modules/basicfuncs/tf-iterate.c:101]: A g_mutex_clear() call may be missing.

Comment (furiel) [reply modules/basicfuncs/tf-iterate.c:101]: Fixed thanks!

**Comment by kira-syslogng** [comment]
Build SUCCESS

**Comment by kira-syslogng** [comment]
Build SUCCESS

**Pull request merged**